[golang] Update to 1.19rc2 (from 1.18.3)

dims opened the pull request:
Go 1.19 GA release is planned for the first week of August:
https://groups.google.com/g/golang-dev/c/4xsD_D5oflI/m/77Ofgt_EAAAJ?utm_medium=email&utm_source=footer

With this PR, we would update to go1.19 rc2 and when the GA is available, we would move to that.

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
For v1.25, Kubernetes will be using golang 1.19, In this PR we update to 1.19rc2 as GA is not yet available.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dims committed:
golang: Update to 1.19rc2 from 1.18.3

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

leilajal commented:
/remove-sig api-machinery

leilajal commented:
/remove-sig api-machinery

liggitt commented:
huh... did 1.19 gofmt make all those comment changes?

liggitt commented:
answering my own question: yes - https://tip.golang.org/doc/go1.19#go-doc

dims commented:
@liggitt ack thanks. I added a commit after running `hack/update-gofmt.sh` with 1.19rc2

liggitt commented:
> @liggitt ack thanks. I added a commit after running `hack/update-gofmt.sh` with 1.19rc2

If previous versions of gofmt leave comments alone, I think we could hoist the result of running 1.19rc2 `gofmt` into a standalone PR and go ahead and merge it ahead of bumping 1.18 → 1.19 if we want to, to keep this one as small as possible

dims commented:
> > @liggitt ack thanks. I added a commit after running `hack/update-gofmt.sh` with 1.19rc2
> 
> If previous versions of gofmt leave comments alone, I think we could hoist the result of running 1.19rc2 `gofmt` into a standalone PR and go ahead and merge it ahead of bumping 1.18 → 1.19 if we want to, to keep this one as small as possible

good point! will do so

dims commented:
/test all

dims committed:
Generate and format files

- Run hack/update-codegen.sh
- Run hack/update-generated-device-plugin.sh
- Run hack/update-generated-protobuf.sh
- Run hack/update-generated-runtime.sh
- Run hack/update-generated-swagger-docs.sh
- Run hack/update-openapi-spec.sh
- Run hack/update-gofmt.sh

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims commented:
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-verify-go-canary

aojea commented:
are we try to get this in 1.25?

dims commented:
/test pull-kubernetes-verify-go-canary

dims commented:
> are we trying to get this in 1.25?

Dunno yet. I am exploring how much work is needed for sure. 

- https://kubernetes.slack.com/archives/C2C40FMNF/p1658173936650579
- https://kubernetes.slack.com/archives/CHGFYJVAN/p1658159216005969

dims commented:
/retest
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-e2e-kind

dims committed:
Try disabling logcheck for golang 1.19rc2

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims commented:
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-unit

dims commented:
/test pull-kubernetes-e2e-gce-storage-slow

dims commented:
/assign @liggitt @thockin

dims commented:
NOTE: see that `pull-kubernetes-verify` is failing but `pull-kubernetes-verify-go-canary` is passing, this is because the verify job is still using go1.18 while the verify-go-canary was bumped to golang 1.19rc2 in https://github.com/kubernetes/test-infra/pull/26860/files

dims commented:
/test pull-kubernetes-verify-go-canary

dims commented:
(simple rebase, no changes!)

dims commented:
/priority important-soon

dims commented:
/test pull-kubernetes-e2e-capz-azure-file

dims commented:
/skip

leilajal commented:
/remove-sig api-machinery

liggitt commented on pkg/fieldpath/fieldpath.go:
```diff
@@ -84,15 +84,15 @@ func ExtractFieldPathAsString(obj interface{}, fieldPath string) (string, error)
-//  - "metadata.labels"               --> ("metadata.labels", "", false)
+//   - "metadata.annotations['myKey']" --> ("metadata.annotations", "myKey", true)
+//   - "metadata.annotations['a[b]c']" --> ("metadata.annotations", "a[b]c", true)
+//   - "metadata.labels[”]"           --> ("metadata.labels", "", true)
```
this is actually wrong... did gofmt make this change?

dims replied:
yes, it did.

Let me replace that example using double quotes `metadata.labels[”]"`

liggitt replied:
but double quotes aren't valid syntax for fieldPath indexing... only single-quotes are :-/

liggitt replied:
is there a way to indent or wrap this so gofmt leaves it alone?

liggitt replied:
on second thought, this might actually be a bug worth reporting to go... indenting and tweaking list bullets is one thing... replacing characters inside godoc strings is much more invasive

liggitt replied:
looks like we want these to be code blocks:

> A code block is a span of indented or blank lines not starting with a bullet list marker or numbered list marker. It is rendered as preformatted text (a <pre> block in HTML).

maybe dropping the leading list markers and putting blank lines between them would make gofmt leave these alone?

liggitt commented on pkg/kubelet/pluginmanager/cache/types.go:
```diff
@@ -20,28 +20,31 @@ package cache
-//                         | Error                                | Error                        |
-//                         |                                      |                              |
-//                         v                                      v                              v
-//                        Out                                    Out                            Out
+//   - +                              +
+//     |                                      |                              |
+//     | Error                                | Error                        |
+//     |                                      |                              |
+//     v                                      v                              v
+//     Out                                    Out                            Out
```
oof... gofmt is not kind to ascii art

dims replied:
fixing. i am not quite sure how this got tripped up.

liggitt commented on staging/src/k8s.io/kubectl/pkg/util/fieldpath/fieldpath.go:
```diff
@@ -86,15 +86,15 @@ func ExtractFieldPathAsString(obj interface{}, fieldPath string) (string, error)
-//  - "metadata.labels"               --> ("metadata.labels", "", false)
+//   - "metadata.annotations['myKey']" --> ("metadata.annotations", "myKey", true)
+//   - "metadata.annotations['a[b]c']" --> ("metadata.annotations", "a[b]c", true)
+//   - "metadata.labels[”]"           --> ("metadata.labels", "", true)
```
same here, this syntax is wrong now

dims replied:
same ... switching to double quotes

liggitt commented:
do we have capability to run a canary integration job on 1.19rc2? I remember the bump to 1.18 hit integration issues once CI moved

dims committed:
Fix quotes that trip up gofmt

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

liggitt commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary

liggitt commented:
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims committed:
Adjust for os/exec changes in 1.19

https://tip.golang.org/doc/go1.19#os-exec-path

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/test pull-kubernetes-e2e-ubuntu-gce-network-policies

dims commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
/skip

dims commented:
@liggitt @thockin this is ready to merge. Please take a look.

dims commented:
also see https://github.com/kubernetes/kubernetes/issues/111396

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -173,21 +174,41 @@ func NewDefaultPluginHandler(validPrefixes []string) *DefaultPluginHandler {
 		path, err := exec.LookPath(fmt.Sprintf("%s-%s", prefix, filename))
-		if err != nil || len(path) == 0 {
-			continue
+		if err != nil {
```
Since 1.19 gives us a specific error indicator for the dot-path lookup, can we use that rather than string checks?

I think that would limit the changes in this file to:
```diff
- if err != nil || len(path) == 0 {
+ if shouldSkipOnLookPathErr(err) || len(path) == 0 {
```

then the <=go1.18 impl is:
```go
// +build !go1.19
...
func shouldSkipOnLookPathErr(err error) bool {
  return err != nil
}
```

and the 1.19+ impl is:
```go
// +build go1.19
...
func shouldSkipOnLookPathErr(err error) bool {
  return err != nil && !errors.Is(err, exec.ErrDot)
}
```

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -173,21 +174,41 @@ func NewDefaultPluginHandler(validPrefixes []string) *DefaultPluginHandler {
+		Args: append([]string{name}, arg...),
+	}
+	if filepath.Base(name) == name {
+		lp, _ := exec.LookPath(name)
+		if lp != "" {
```
could use the same error checker helper here just to make sure we're not getting some other error we actually want to surface:

```go
lp, err := exec.LookPath(name)
if lp != "" && !shouldSkipOnLookPathErr(err) {
  cmd.Path = lp
```

dims commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

dims commented:
@MadhavJivrajani @liggitt please see the very last commit in this one, has the changes for Lookup/osexec stuff

liggitt commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -173,21 +174,37 @@ func NewDefaultPluginHandler(validPrefixes []string) *DefaultPluginHandler {
+			// Update cmd.Path even if err is non-nil.
+			// If err is ErrDot (especially on Windows), lp may include a resolved
+			// extension (like .exe or .bat) that should be preserved.
+			cmd.Path = lp
```
Were you able to try this on a windows machine (or ask someone from sig-windows to)? Do we also need to replace `Args[0]` with the fully resolved binary? (I don't know enough about how Command works on Windows to know)

dims replied:
@liggitt i am fairly sure this is fine as it is from https://github.com/golang/go/blob/14abe8aa7324bdf0e09e1dfebfb3519cc30f4918/src/os/exec/exec.go#L273-L278 :)

liggitt replied:
ha, ok

liggitt commented:
> @MadhavJivrajani @liggitt please see the very last commit in this one, has the changes for Lookup/osexec stuff

Thanks, all the changes I understand lgtm, had one question about the Args[0] use of the passed-in name

liggitt commented on hack/lib/golang.sh:
```diff
@@ -482,6 +482,7 @@ EOF
   local go_version
   IFS=" " read -ra go_version <<< "$(GOFLAGS='' go version)"
   local minimum_go_version
+  # TODO(dims): Need to switch this to 1.19 once we update images to newer go version
```
once this merges and CI switches to 1.19rc2, will people with a local go1.18 be generating/formatting changes that will fail CI?

dims replied:
Yes, there is one file that has trouble, this is what happens when i run update-gofmt.sh using golang 1.18 (with this branch)

```
diff --git staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go
index a5b01802cb95..b468b6d3b0d1 100644
--- staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go
+++ staging/src/k8s.io/apiserver/pkg/util/flowcontrol/patch_test.go
@@ -59,10 +59,10 @@ func Test_configController_generatePatchBytes(t *testing.T) {
                {
                        name: "check when message has a whitespace character that can be escaped",
                        condition: flowcontrol.FlowSchemaCondition{
-                               Type:               flowcontrol.FlowSchemaConditionDangling,
-                               Status:             flowcontrol.ConditionTrue,
-                               Reason:             "test reason",
-                               Message:            `test               none`,
+                               Type:   flowcontrol.FlowSchemaConditionDangling,
+                               Status: flowcontrol.ConditionTrue,
+                               Reason: "test reason",
+                               Message: `test          none`,
                                LastTransitionTime: metav1.NewTime(now),
                        },
                        want: []byte(fmt.Sprintf(`{"status":{"conditions":[{"type":"Dangling","status":"True","lastTransitionTime":"%s","reason":"test reason","message":"test \t\tnone"}]}}`, now.Format(time.RFC3339))),

```

The `Message` has a horizontal tab (hex 0x09) in the string and how that is handled is different between the go versions

liggitt replied:
interesting... 1.19 gofmt and 1.18 gofmt dueling seems like a bug we should report... my understanding is that gofmt tries to only make changes a previous version would leave alone

as a workaround for us, you could put a blank line before and after the `Message` line and I bet both go1.19 and go1.18 would format it identically

liggitt replied:
> interesting... 1.19 gofmt and 1.18 gofmt dueling seems like a bug we should report

well, looks intentional, given https://github.com/golang/go/issues/51910 / https://github.com/golang/go/commit/bf68170c638e7e69bedcc64fadfd83354fd06c10

liggitt replied:
the workaround to avoid adding formatting friction to ~everyone's PRs the week of code freeze is probably worthwhile

dims replied:
`blank line before and after the Message line` did not help @liggitt 

Can i switch this to "test  \u0009none" (2 spaces and then a tab are in between `"test"` and `"one"`

liggitt replied:
> Can i switch this to "test \u0009none" (2 spaces and then a tab are in between "test" and "one"

yup, that'd work too

dims replied:
Done!

dims commented:
@liggitt ack thanks, will reorder stuff later today!

dims committed:
fix patch_test for gofmt issue

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims commented:
@liggitt @thockin reordered/squashed commits as requested in https://github.com/kubernetes/kubernetes/pull/111254#issuecomment-1194207441

will ping when this CI goes green. (used diff to make sure i did not make a mistake in the reorder/squash process. rebased to master as well)

liggitt commented:
Changes lgtm (and fwiw, I checked out the PR and ran gofmt with go1.18 and there were no changes, which is great)

Sweeping what 1.19 gofmt "autocorrected" in our doc strings, there's a fair number of places we had weird indents that gofmt is turning into new paragraphs, etc. I think this PR should go ahead and merge so we can get go1.19 signal, but a followup to sweep the doc auto format changes in this PR and fix up ones that look wrong would be good.

/lgtm
/approve
/hold to make sure canary presubmits are green

mechanically, once canary presubmits are green, I guess we would:
1. update real unit/integration/verify presubmits to use the go1.19rc2 images (which will make all other PRs to master start failing on format/verify errors)
2. retest this PR to get it to pass blocking presubmits and merge it

A heads-up that this is coming is probably worth communicating to dev@kubernetes.io ... the format changes on generated files is unfortunate, as it will make a ton of PRs require rebase

liggitt commented:
/test pull-kubernetes-integration-go-canary
/test pull-kubernetes-unit-go-canary
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

klueska commented:
/approve

dims commented:
/test pull-kubernetes-verify 
/test pull-kubernetes-verify-go-canary

liggitt commented:
pull-kubernetes-verify-go-canary passed

looks like all the verify failures were formatting ones which will resolve when we bump test-infra

dims commented:
@liggitt yep that landed in https://github.com/kubernetes/test-infra/pull/26920 and i am retesting the 2 jobs after that.

leilajal commented:
/remove-sig api-machinery

dims commented:
/hold cancel

lavalamp commented:
This is causing an annoying gofmt change: https://github.com/kubernetes-sigs/structured-merge-diff/pull/220/files#diff-054f947623025b6f29ed3f11f2849794e09e83d9a31927598d4c1f059e7c4fe1

cc @alexzielenski @apelisse

liggitt commented:
xref https://tip.golang.org/doc/go1.19#go-doc - As part of this change, [gofmt](https://tip.golang.org/cmd/gofmt) now reformats doc comments to make their rendered meaning clearer.

I'd recommend committing the go1.19 version of the doc, which go1.18 gofmt will leave alone

apelisse commented:
Yeah, nested lists aren't supported in comments by gofmt which is unfortunate and clearly changes the meaning of some comments.

alexzielenski commented:
filed an issue with golang https://github.com/golang/go/issues/54129

lavalamp commented:
There's an example at the very bottom here: https://tip.golang.org/doc/comment

I guess they will probably close the issue :(
